Compilation with Ogre & OIS - Windows - Basic Scene, Entity and Keyboard Handling

Arignir opened the pull request:
Changelog:

* Modified CMakeLists.txt : addded compilation and link with Ogre and OIS
* Added code to open a window (1920 * 1080, fullscreen)
* Added main class `Game` as a singleton. Contains the Ogre main pointer : `Ogre::Root` and some other main pointers (Window, InputManager, Renderer etc.).
* Added a `Renderer` class to handle different scenes and camera.
* Added a `Scene` class to handle differents scenes, populate / unpopulate them, register key callbacks etc.
* Added a `Entity` class to wrap the couple Ogre::Entity and Ogre::SceneNode.
* Added an `Input` class to wrap all input methods (Keyboard, mouse, controller etc.)
* Added a `Keyboard` class to wrap keyboards input. It's a singleton.
* Added a demo scene to show off what's done so far. There is an Ogre head on a plane. You can move the camera with ZQSD (Azerty keyboard), and rotate the camera with A and E. Press Space to turn off the light.

Related issues:

 https://github.com/Gaspard--/Swords-Scrolls-and-Knuckles/issues/4
 https://github.com/Gaspard--/Swords-Scrolls-and-Knuckles/issues/3
 https://github.com/Gaspard--/Swords-Scrolls-and-Knuckles/issues/2

Gaspard-- committed:
Merge pull request #1 from Arignir/master

Updated README.md

Arignir committed:
Updated README.md

Arignir committed:
Added basic CMAKE compilation, linking with Ogre and OIS.

Arignir committed:
Updated README.md and added compilation informations.

Arignir committed:
Updated .gitignore to match CMake temporary files.

Gaspard-- committed:
Merge pull request #6 from Arignir/master

CMake compilation, linking with Ogre3D.

Arignir committed:
Fixed spelling mistake in ASCII art and removed {EPITECH.} header.

Gaspard-- committed:
Update README.md

Fixed typo.

Gaspard-- committed:
Update README.md

Corrected another typo.

Gaspard-- committed:
Update README.md

Fixed another typo, this is getting tedious :P

Arignir committed:
Added App class

Arignir committed:
Moved the output binary one directory up

Arignir committed:
Updated README.md

Arignir committed:
Made a very basic Scene system. Added some Ogre resources (mesh, material and textures).

Arignir committed:
Revamped Renderer::switchScene(), removed unique_ptr on Ogre objects

Arignir committed:
Removed useless scope

Arignir committed:
Made Game a singleton. Beautified demo scene

Arignir committed:
Added doc here and there

Arignir committed:
Merge branch 'master' of https://github.com/Gaspard--/Swords-Scrolls-and-Knuckles

Arignir committed:
Removed useless {EPITECH.} headers

Gaspard-- committed:
Create CodingStyle.md

HAHAHAHHAHAHHA

Arignir committed:
Added a very basic Entity system to encapsulate Ogre::Entity and Ogre::SceneNode

Gaspard-- committed:
Update CodingStyle.md

Tried to make guide more complete. Corrected typos and made more readable.

Gaspard-- committed:
Update CodingStyle.md

Added a few guides for classes.

Arignir committed:
Added Keyboard handling

Arignir committed:
Added buffered input. Documented here and there

Arignir committed:
Beautified Game

Arignir committed:
Added an update() function in Scene. Added some movements / rotation with the camera in the Demo Scene.

Arignir committed:
Qualified static char const * as constexpr

Arignir committed:
Merge branch 'master' of https://github.com/Gaspard--/Swords-Scrolls-and-Knuckles

Gaspard-- commented:
I'm going to report problemes as I go through the sources:
- DemoScene overrides load & update from scene, put lacks the `override` keyword on those functions
- `Entity` has a virtual destructor, even if it doesn't look like it should be inherited.
- Most of `Entity`'s constructors could be consexpr (not really a problem, i just need a rational)
- Why do `Entity`'s getters return pointers? (not even references to pointers to be able to set them)
- `Game` inherits from 2 classes why? (there are other details in Game to handle too)
- `Game::setCurrentScene` takes a pointer. Why?
- `Input::oisobject` isn't in camelCase :P 
- Why does `Input` inherit? Shouldn't it be `Keyboard` ?
- `Input` doesn't seem to adhere to RAII
- `Keyboard` however, seems to be inheriting 2 functions not marked override.
- `KeyboardKey`'s constructor should be constexpr.
- `Renderer` has 4 variables preceded by an underscore.
- `Scene` has 1
- Why does `Scene` have a reference on a `Renderer` ?
- `Game` does not seem to respect RAII.
- `Scene` has empty virtual functions, maybe they should be = 0?

Finally: we seem to have many singletons, are they really all necessary?

Gaspard-- committed:
Update CodingStyle.md

Added nuance to operator overloading: not everything is always as simple as the given case.

Arignir committed:
Fixed coding style errors

Arignir commented:
Hi.

Thanks for the review.

`DemoScene overrides load & update from scene, put lacks the override keyword on those functions`

Fixed.

`Entity has a virtual destructor, even if it doesn't look like it should be inherited.`

It's meaned to be inherited by all sorts of subentities (EntityZombie, EntityPlayer etc. to share common attributes (collisions, gravity, health etc.)

`Most of Entity's constructors could be consexpr (not really a problem, i just need a rational)`

Fixed.

`Why do Entity's getters return pointers? (not even references to pointers to be able to set them)`

Can't return a reference, as there is no guarantee the pointers are valid (Entity could be default initialised). I've set the return value as reference to pointers while waiting for a better solution.

`Game inherits from 2 classes why? (there are other details in Game to handle too)`

Ogre's tutorials main class inherits these two classes. Main interest is to override some protected functions (`frameRenderingQueue` and `windowClosed`, to build up the main loop and destroy everything correctly when the window is closing).

`Game::setCurrentScene takes a pointer. Why?`

Woops, forgot to delete the prototype of this function, my bad. It is not implemented in the cpp file. It's equivalent is `Renderer::switchScene(std::unique_ptr<Scene> &&)`.

`Input::oisobject isn't in camelCase :P`

Fixed

`Why does Input inherit? Shouldn't it be Keyboard ?`

Shit, a copy / paste miss. Fixed.

`Input doesn't seem to adhere to RAII`

I think you're talking about the `destroy()` function. The problem is that we need a pointer on `OIS::InputManager *rm` to delete the content, and destructors doesn't take any parameter. So we could store the pointer internally, but Keyboard is a singleton, and so it will likely be deleted after the InputManager. Any idea here ?

`Keyboard however, seems to be inheriting 2 functions not marked override.`

Fixed.

`KeyboardKey's constructor should be constexpr.`

Fixed.

`Renderer has 4 variables preceded by an underscore.` and `Scene has 1`

Woops, forgot that. Fixed.

`Why does Scene have a reference on a Renderer ?`

Can't remember. I removed it.

`Game does not seem to respect RAII.`

Can you be a bit more specific please ?
I've added some `if` in the destructor in case `window` equals `nullptr`. 

`Scene has empty virtual functions, maybe they should be = 0?`

It's a default implementation. Some scene may not need an unload / update function. Forcing their implementation isn't really needed in my opinion. It doesn't change anything, so why ?

`Finally: we seem to have many singletons, are they really all necessary?`

They are three singleton:
* Game::getGame(): It's used every where, only one instance, from the beginning to the end.
* Keyboard::getKeyboard(): I can put this as a member object of Game, but it looks like pretty much the same for me
* Keyboard::getKeys() : This one is only sugar on getKeyboard().

Gaspard-- commented:
`It's meaned to be inherited by all sorts of subentities (EntityZombie, EntityPlayer etc. to share common attributes (collisions, gravity, health etc.)`
Composition could be used instead, which is much cleaner in my opinion.

Thanks for the clarification about Scene. The introduced inheritance is accepted only because it won't affect performance much. However that doesn't apply to entity.

`I think you're talking about the destroy() function. The problem is that we need a pointer on OIS::InputManager *rm to delete the content, and destructors doesn't take any parameter. So we could store the pointer internally, but Keyboard is a singleton, and so it will likely be deleted after the InputManager. Any idea here ?`

This looks like a logical problem. I'm going to look into it.

Gaspard-- commented:
Latest version does not compile:
```
/home/kellen_j/Documents/Swords-Scrolls-and-Knuckles/source/Game.cpp: In member function ‘void Game::setupResources()’:
/home/kellen_j/Documents/Swords-Scrolls-and-Knuckles/source/Game.cpp:38:29: error: ‘class Ogre::ConfigFile’ has no member named ‘getSettingsBySection’; did you mean ‘SettingsBySection’?
   auto const &sections = cf.getSettingsBySection();
                             ^~~~~~~~~~~~~~~~~~~~
/home/kellen_j/Documents/Swords-Scrolls-and-Knuckles/source/Game.cpp:39:25: error: unable to deduce ‘auto&&’ from ‘sections’
   for (auto const &it : sections) {
                         ^~~~~~~~
CMakeFiles/ssk.dir/build.make:110: recipe for target 'CMakeFiles/ssk.dir/source/Game.cpp.o' failed

```

Arignir committed:
Fixed GCC warning in Game's constructor

Arignir committed:
Switched to Ogre 1.9

Commit:
cmake searches for lib and generates plugins.config

Commit:
removed useless file

Commit:
updated to conform to code review

Arignir committed:
Merge pull request #1 from Gaspard--/vect

Updated cmake

Gaspard-- commented on include/Input.hpp:
```diff
@@ -0,0 +1,42 @@
+
+    void init(OIS::Type t, OIS::InputManager *im)
+    {
+      oisObject =static_cast<T *>(im->createInputObject(t, true));
```
missing space after =

Gaspard-- commented on include/Input.hpp:
```diff
@@ -0,0 +1,42 @@
+    void init(OIS::Type t, OIS::InputManager *im)
+    {
+      oisObject =static_cast<T *>(im->createInputObject(t, true));
+      oisObject->setEventCallback(this);
```
You seem to have a better idea for this part.

Gaspard-- commented on include/Renderer.hpp:
```diff
@@ -0,0 +1,45 @@
+class Renderer
+{
+  private:
+    Ogre::SceneManager *_scenemgr;
```
We still have variables with underscores here.

Gaspard-- commented on include/Game.hpp:
```diff
@@ -0,0 +1,83 @@
+class Game : public Ogre::WindowEventListener, public Ogre::FrameListener
+{
+    /// Singleton instance
+    static Game gameInstance;
```
Still have to do something about this singleton.

Arignir replied:
Can you be more specific ? I can't remember where we ended this conversation.

Gaspard-- replied:
It should be passed as parameter to all functions that require it.

Gaspard-- commented on include/Entity.hpp:
```diff
@@ -0,0 +1,57 @@
+      , sceneNode(std::move(e.sceneNode))
+    {
+      e.ogreEntity = nullptr;
+      e.sceneNode = nullptr;
```
Since the destructor does not act on the pointers of Entity, why the = nullptr?
(note that using a class after it has moved is not defined after the standart)

Arignir replied:
You're right, they are useless.

Gaspard-- commented on include/Entity.hpp:
```diff
@@ -0,0 +1,57 @@
+
+    Entity &operator=(Entity const &e) = default;
+    Entity &operator=(Entity &&e) = default;
+    virtual ~Entity(void) = default;
```
If we use composition, the destructor shouldn't be virtual.

Gaspard-- commented on include/DemoScene.hpp:
```diff
@@ -0,0 +1,34 @@
+# include "Scene.hpp"
+# include "Entity.hpp"
+
+class Game;
```
Why is game forward declared if it isn't used?

Gaspard-- commented on include/Scene.hpp:
```diff
@@ -0,0 +1,29 @@
+# define SCENE_HPP
+
+class Game;
+class Renderer;
```
Renderer & Game both seem unused.

Gaspard-- commented on include/Renderer.hpp:
```diff
@@ -0,0 +1,45 @@
+# include <OgreViewport.h>
+# include "Scene.hpp"
+
+class Game;
```
Game seems unused.

Gaspard-- commented on include/Game.hpp:
```diff
@@ -0,0 +1,83 @@
+
+    /// Ogre config plugins & resource path
+    static constexpr char const *PLUGINS_CONFIG_PATH = "resources/cfg/plugins.cfg";
+    static constexpr char const *RESOURCES_CONFIG_PATH = "resources/cfg/resources.cfg";
```
Could these be private?

Gaspard-- commented on include/Entity.hpp:
```diff
@@ -0,0 +1,57 @@
+
+    /// Get the ogre's version of this entity.
+    Ogre::Entity *&getOgre(void);
+    Ogre::Entity * const &getOgre(void) const;
```
ref useless, return a simple copy.

Gaspard-- commented on include/Entity.hpp:
```diff
@@ -0,0 +1,57 @@
+
+    /// Get the node attached to this entity.
+    Ogre::SceneNode *&getNode(void);
+    Ogre::SceneNode * const &getNode(void) const;
```
Same as above.

Gaspard-- commented on include/Vect.hpp:
```diff
@@ -0,0 +1,326 @@
+#ifndef VECT_HPP_
```
WHo THE FuCK WRoTE ThIS PIeCE oF SHiT? i SURe HOPe He LeARNs hE lEarns C++14! /s

Arignir replied:
I let you fix this one :)

Gaspard-- replied:
:') I'll open an issue, this isn't top priority anyway.

Gaspard-- reviewed (changes requested):
I did some reviewing and trolling :)

Arignir committed:
Updated to conform to comments & coding style

Arignir committed:
Merge branch 'master' of github.com:Arignir/Swords-Scrolls-and-Knuckles

Arignir committed:
Edited CMakeLists.txt to make it more portable

Arignir commented:
I've updated everything that you pointed out (i hope i didn't missed one).
My idea about Input would not work, so i'm going to stick with what's present already.

Also, i've updated `CMakeLists.txt` because the modifications you made about `Plugins.cfg` didn't work on my computer. Can you please tell me if this version works on both our two computers ?

Is there anything else to edit ?

Arignir committed:
Merge branch 'master' of https://github.com/Gaspard--/Swords-Scrolls-and-Knuckles

Gaspard-- commented:
I'll check this out.

Arignir committed:
Made CMakeLists.txt even more portable

Gaspard-- commented on include/Game.hpp:
```diff
@@ -0,0 +1,78 @@
+    ~Game(void);
+
+    /// Singleton getter
+    static Game &getGame(void);
```
I think you forgot this.

Gaspard-- commented on source/Game.cpp:
```diff
@@ -0,0 +1,159 @@
+void Game::setupRenderSystem(void) {
+  // Look for the OpenGL render system
+  auto const &rlist = root.getAvailableRenderers();
+
```
Is something preventing rlist from being inlined?

Gaspard-- commented on source/Game.cpp:
```diff
@@ -0,0 +1,159 @@
+
+  // Update the current scene's logic
+  if (renderer->getScene()) {
+    go_on &= renderer->getScene()->update(*this);
```
this will not short-circuit (I don't think you want that so isok, just telling you)

Gaspard-- commented on source/Game.cpp:
```diff
@@ -0,0 +1,159 @@
+
+// Public functions
+
+void Game::setup(void) {
```
Is this needed if Game is no longer a singleton?

Arignir replied:
You mean, all `Game::setup(void)` content in the constructor ?

Gaspard-- replied:
Yes.

Gaspard-- commented on source/Renderer.cpp:
```diff
@@ -0,0 +1,52 @@
+  return (*camera);
+}
+
+std::unique_ptr<Scene> &Renderer::getScene(void) {
```
Do you want this unique_ptr to be modifiable? (seems a weird choice)

Arignir committed:
Fixed code-review problems

Gaspard-- commented:
Good for me!
